src/H5Dcontig.c

```diff
@@ -562,33 +574,160 @@ H5D__contig_is_data_cached(const H5D_shared_t *shared_dset)
+
+    int sf_ndims; /* The number of dimensions of the file dataspace (signed) */
+
     htri_t use_selection_io = FALSE;   /* Whether to use selection I/O */
-    htri_t ret_value        = SUCCEED; /* Return value */
+    herr_t ret_value        = SUCCEED; /* Return value */
 
     FUNC_ENTER_PACKAGE
 
-    io_info->store->contig.dset_addr = io_info->dset->shared->layout.storage.u.contig.addr;
-    io_info->store->contig.dset_size = io_info->dset->shared->layout.storage.u.contig.size;
+    dinfo->store->contig.dset_addr = dataset->shared->layout.storage.u.contig.addr;
+    dinfo->store->contig.dset_size = dataset->shared->layout.storage.u.contig.size;
+
+    /* Initialize piece info */
+    dinfo->layout_io_info.contig_piece_info = NULL;
+
+    /* Get layout for dataset */
+    dinfo->layout = &(dataset->shared->layout);
+
+    /* Get dim number and dimensionality for each dataspace */
+    if ((sf_ndims = H5S_GET_EXTENT_NDIMS(dinfo->file_space)) < 0)
+        HGOTO_ERROR(H5E_DATASPACE, H5E_CANTGET, FAIL, "unable to get dimension number")
+
+    /* Normalize hyperslab selections by adjusting them by the offset */
+    /* (It might be worthwhile to normalize both the file and memory dataspaces
+     * before any (contiguous, chunked, etc) file I/O operation, in order to
+     * speed up hyperslab calculations by removing the extra checks and/or
+     * additions involving the offset and the hyperslab selection -QAK)
+     */
+    if ((file_space_normalized = H5S_hyper_normalize_offset(dinfo->file_space, old_offset)) < 0)
+        HGOTO_ERROR(H5E_DATASET, H5E_BADSELECT, FAIL, "unable to normalize dataspace by offset")
+
+    /* if selected elements exist */
+    if (dinfo->nelmts) {
+        int               u;
+        H5D_piece_info_t *new_piece_info; /* piece information to insert into skip list */
+
+        /* Get copy of dset file_space, so it can be changed temporarily
+         * purpose
+         * This tmp_fspace allows multiple write before close dset */
+        H5S_t *tmp_fspace; /* Temporary file dataspace */
+
+        /* Create "temporary" chunk for selection operations (copy file space) */
+        if (NULL == (tmp_fspace = H5S_copy(dinfo->file_space, TRUE, FALSE)))
+            HGOTO_ERROR(H5E_DATASPACE, H5E_CANTCOPY, FAIL, "unable to copy memory space")
+
+        /* Add temporary chunk to the list of pieces */
+        /* collect piece_info into Skip List */
+        /* Allocate the file & memory chunk information */
+        if (NULL == (new_piece_info = H5FL_MALLOC(H5D_piece_info_t))) {
+            (void)H5S_close(tmp_fspace);
+            HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, FAIL, "can't allocate chunk info")
+        } /* end if */
+
+        /* Set the piece index */
+        new_piece_info->index = 0;
+
+        /* Set the file chunk dataspace */
+        new_piece_info->fspace        = tmp_fspace;
+        new_piece_info->fspace_shared = FALSE;
```

Comment:
Does setting fspace_shared to FALSE here interact with having called H5S_copy above with TRUE for the share_selection parameter?

Reply:
This *should* be ok because H5S__hyper_copy() keeps a reference count on the span list (and H5S__point_copy() just always copies the point list). If either space's selection is modified it could cause problems, but that shouldn't happen. I'm open to arguments that it should be deep copied anyways (pass share_selection as FALSE) to avoid this non-obvious potential fragility in the code, at a small performance hit.

Reply:
I see that this pattern has been used elsewhere in existing code, so it's probably fine even though it kind of jumped out as possibly problematic right away.